### PR TITLE
Support universal colors and assets

### LINKED ISSUE #68: Support for Universal Colors
I was curious about the validation that ensures there is a corresponding dark mode color for every light mode color. In some cases, a design system might only have a single color that is used across both the light and dark color palette.

Consider something like this:
image

What do you think about adding a configuration option to opt out of that warning? Then the xcode exporter step can check if only a light color exists, it turns it into a color set with the `None` appearances option. This color will then not respond to appearance changes.

### PR DESCRIPTION
**Pull request by JonLz**
Resolves #68 

Let me know if there is any fix required for android. I'm not as familiar with how the color specifications work on that platform. I think it can now generate a dark colors file with less values than the light colors file. Is that an issue? For the time being, I've added in the same warning used in iOS. Similarly, I was not able to test illustration exports but let me know if you see any issues there.

### CODE REVIEW

**Comment by subdan** [comment]
Thanks for the PR! I will check it by the end of the week.

File: Sources/FigmaExport/Subcommands/ExportImages.swift [hunk 1]
```diff
@@ -57,7 +57,10 @@ extension FigmaExportCommand {
                 nameReplaceRegexp: params.common?.images?.nameReplaceRegexp,
                 nameStyle: params.ios?.images.nameStyle
             )
-            let images = try processor.process(light: imagesTuple.light, dark: imagesTuple.dark).get()
+            let images = try processor.process(light: imagesTuple.light, dark: imagesTuple.dark)
```
Comment (subdan) [inline Sources/FigmaExport/Subcommands/ExportImages.swift:60]: Delete unnecessary `try` keyword.

File: README.md [hunk 2]
```diff
@@ -352,7 +352,7 @@ If a color, icon or image is unique for iOS or Android platform, it should conta
 
 For `figma-export colors`
 
-If you support dark mode your figma project must contains two files. One should contains a dark color palette, and the another light color palette. Names and number of the colors must matches.
+If you support dark mode your figma project must contains two files. One should contains a dark color palette, and the another light color palette. The light color palette may contain more colors than the dark color palette. If a light-only color is present, it will be considered a universal color for the iOS color palette. Names of the dark colors must match the light colors.
```
Comment (subdan) [inline README.md:355]: a universal → as universal

File: README.md [hunk 3]
```diff
@@ -369,7 +369,7 @@ For `figma-export images`
 
 Your Figma file should contains a frame with `Illustrations` name which contains components for each illustration. You may change a frame name in a [config](Config.md) file by setting `common.images.figmaFrameName` property.
 
-If you support dark mode you must have two Figma files.
+If you support dark mode you must have two Figma files. The rules for these two files follow the same rules as the color files specified above.
```
Comment (subdan) [inline README.md:372]: Replace
`the same rules as the color files specified above.`
with
`the same rules as described above for colors.`

File: Sources/FigmaExport/Subcommands/ExportColors.swift [hunk 4]
```diff
@@ -36,10 +36,13 @@ extension FigmaExportCommand {
...
                 )
-                let colorPairs = try processor.process(light: colors.light, dark: colors.dark).get()
+                let colorPairs = processor.process(light: colors.light, dark: colors.dark)
+                if let warning = colorPairs.warning?.errorDescription {
+                    logger.warning("\(warning)")
+                }
 
                 logger.info("Exporting colors to Xcode project...")
-                try exportXcodeColors(colorPairs: colorPairs, iosParams: ios, logger: logger)
+                try exportXcodeColors(colorPairs: try colorPairs.get(), iosParams: ios, logger: logger)
```
Comment (subdan) [inline Sources/FigmaExport/Subcommands/ExportColors.swift:45]: Delete unnecessary `try` keyword.

Comment (subdan) [reply Sources/FigmaExport/Subcommands/ExportColors.swift:45]: I mean the second one.

File: Sources/FigmaExport/Subcommands/ExportImages.swift [hunk 5]
```diff
@@ -112,13 +115,16 @@ extension FigmaExportCommand {
...
+                logger.warning("\(warning)")
+            }
             
             switch androidImages.format {
             case .svg:
-                try exportAndroidSVGImages(images: images, params: params, logger: logger)
+                try exportAndroidSVGImages(images: try images.get(), params: params, logger: logger)
             case .png, .webp:
-                try exportAndroidRasterImages(images: images, params: params, logger: logger)
+                try exportAndroidRasterImages(images: try images.get(), params: params, logger: logger)
```
Comment (subdan) [inline Sources/FigmaExport/Subcommands/ExportImages.swift:127]: Delete unnecessary second `try` keyword here and above.

**Comment by JonLz** [comment]
Thank you for reviewing the PR. I've added your suggestions.

**Comment by JonLz** [comment]
Rebased from the latest master and updated the README conflicts.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 6]
```diff
@@ -29,10 +31,31 @@ public extension AssetNameProcessable {
     }
 }
 
+public struct AssetResult<Success, Error> {
+    public var result: Result<Success, Swift.Error>
+    public var warning: AssetsValidatorWarning?
+
+    public func get() throws -> Success {
+        return try result.get()
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:39]: Omit needless `return` keyword here and below.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 7]
```diff
@@ -29,10 +31,31 @@ public extension AssetNameProcessable {
     }
 }
 
+public struct AssetResult<Success, Error> {
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:34]: Extract the struct to a new file.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 8]
```diff
@@ -270,30 +299,40 @@ public extension AssetsProcessable {
             .filter { $0.platform == platform || $0.platform == nil }
             .sorted { $0.name < $1.name }
 
-        let darkColors = darkSet
-            .filter { $0.platform == platform || $0.platform == nil }
-            .sorted { $0.name < $1.name }
+        // After validations, only those dark colors in the light color palette are allowed
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:302]: This function works for any type of assets: colors, images. Let’s rename all “color” word to “asset” word.
lightColors → lightAssets, darkColorMap → darkAssetMap, etc.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 9]
```diff
@@ -270,30 +299,40 @@ public extension AssetsProcessable {
...
+            .map { lightColor in darkColorMap[lightColor.name] }
 
         let zipResult = zip(lightColors, darkColors)
 
         return zipResult
             .map { lightAsset, darkAsset in
+                if let darkAsset = darkAsset {
+                    return AssetPair(light: processedAssetName(lightAsset), dark: processedAssetName(darkAsset))
+                } else {
+                    return AssetPair(light: processedAssetName(lightAsset), dark: nil)
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:316]: Get rid of code duplication by using `map`:
```swift
AssetPair(
    light: processedAssetName(lightAsset),
    dark: darkAsset.map { processedAssetName($0) }
)
```

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 10]
```diff
@@ -270,30 +299,40 @@ public extension AssetsProcessable {
...
 
-        let darkColors = darkSet
-            .filter { $0.platform == platform || $0.platform == nil }
-            .sorted { $0.name < $1.name }
+        // After validations, only those dark colors in the light color palette are allowed
+        // However the dark array may be smaller than the light array
+        // Create a same size array of dark colors so we can zip in the next step
+        let darkColorMap: [String: AssetType] = darkSet.reduce(into: [:]) { $0[$1.name] = $1 }
+        let darkColors = lightColors
+            .map { lightColor in darkColorMap[lightColor.name] }
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:307]: There is no need to move .map to the next line.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 11]
```diff
@@ -270,30 +299,40 @@ public extension AssetsProcessable {
...
-                    newLightAsset.name = self.replace(newLightAsset.name, matchRegExp: regexp, replaceRegExp: replaceRegExp)
-                    newDarkAsset.name = self.replace(newDarkAsset.name, matchRegExp: regexp, replaceRegExp: replaceRegExp)
-                }
-                
-                if let style = nameStyle {
-                    newLightAsset.name = self.normalizeName(newLightAsset.name, style: style)
-                    newDarkAsset.name = self.normalizeName(newDarkAsset.name, style: style)
-                }
+        if let replaceRegExp = nameReplaceRegexp, let regexp = nameValidateRegexp {
+            newAsset.name = self.replace(newAsset.name, matchRegExp: regexp, replaceRegExp: replaceRegExp)
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:328]: Omit needless `self` keyword here and below.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift [hunk 12]
```diff
@@ -270,30 +299,40 @@ public extension AssetsProcessable {
...
+                }
+            }
+    }
 
-                var newLightAsset = lightAsset
-                var newDarkAsset = darkAsset
+    /// Runs the name replacement and name validation regexps, and name styles, if they are defined
+    /// - Returns:
+    ///   - `AssetType` with a processed name
+    private func processedAssetName(_ asset: AssetType) -> AssetType {
```
Comment (subdan) [inline Sources/FigmaExportCore/Processor/AssetsProcessor.swift:324]: It‘s a good idea to extract name processing code to a new function. Could you please refactor code and call this function in the `makeSortedAssetPairs` and `validateAndProcess` function to get rid of code duplication?

**Review by subdan — CHANGES_REQUESTED** [review]
Read my comments

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift (hunk 8 above)
Comment (JonLz) [reply Sources/FigmaExportCore/Processor/AssetsProcessor.swift]: Good idea. Done.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift (hunk 9 above)
Comment (JonLz) [reply Sources/FigmaExportCore/Processor/AssetsProcessor.swift]: Much cleaner. Thanks.

File: Sources/FigmaExportCore/Processor/AssetsProcessor.swift (hunk 12 above)
Comment (JonLz) [reply Sources/FigmaExportCore/Processor/AssetsProcessor.swift:277]: I didn't catch those myself, good idea. Done.

**Comment by JonLz** [comment]
Thanks for your comments, I think the processing code is cleaner now. Please take a look at the latest commit. Thanks.

**Comment by subdan** [comment]
If I remove a dark illustration I get the following error:
```Error: ❌ Asset with name img_zero_error have different description. In dark theme «», in light theme «android»```

Open `AssetProcessor.swift` file and find `4. descriptionMismatch` string. When there is no dark variant of image `dark` is equal to `nil` and error occurs.

```swift
if let platform = asset.platform {
let dark = darkSet.first(where: { $0.name == asset.name })
if dark?.platform != platform { // <<<<<<—————— Bug here
    errors.all.append(...)
}
```

I've fixed this issue. You can copy the code below.
```swift
// 4. descriptionMismatch
lightSet.forEach { asset in
    if
        let platform = asset.platform,
        let dark = darkSet.first(where: { $0.name == asset.name }),
        dark.platform != platform {
        
        let error = AssetsValidatorError.descriptionMismatch(
            assetName: asset.name,
            light: platform.rawValue,
            dark: dark.platform?.rawValue ?? "")
        
        errors.all.append(error)
    }
}
```

**Review by subdan — APPROVED** [review]

**Comment by subdan** [comment]
@JonLz Thanks for the PR! Great job! I will publish a new version by the end of the week.

**Pull request merged**